File: java_projects/glue_code/commons-codec/src/main/java/org/apache/commons/codec/digest/B64.java
package org.apache.commons.codec.digest;

import java.util.Random;
import org.apache.commons.codec.ContextInitializer;
import org.graalvm.polyglot.Value;

class B64 {
  static final String B64T_STRING =
      "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
      static final char[] B64T_ARRAY = B64T_STRING.toCharArray();
  private static Value clz = ContextInitializer.getPythonClass("/digest/B64.py", "B64");
  private Value obj;

  public B64(Value obj) {
    this.obj = obj;
  }

  public Value getPythonObject() {
    return obj;
  }

  static String getRandomSalt(final int num, final Random random) {
    //
    // final StringBuilder saltString = new StringBuilder(num);
    // for (int i = 1; i <= num; i++) {
    // saltString.append(B64T_STRING.charAt(random.nextInt(B64T_STRING.length())));
    // }
    // return saltString.toString();
    //

    // TODO: Check the type mapping below!
    return clz.invokeMember("getRandomSalt", num, random).as(String.class);
  }

  static void b64from24bit(
      final byte b2, final byte b1, final byte b0, final int outLen, final StringBuilder buffer) {
    //
    // int w = ((b2 << 16) & 0x00ffffff) | ((b1 << 8) & 0x00ffff) | (b0 & 0xff);
    // int n = outLen;
    // while (n-- > 0) {
    // buffer.append(B64T_ARRAY[w & 0x3f]);
    // w >>= 6;
    // }
    //

    clz.invokeMember("b64from24bit", b2, b1, b0, outLen, buffer);
  }
}

File: data/recomposed_projects/llama-3-70b-instruct/body/commons-codec/src/main/org/apache/commons/codec/digest/B64.py
from __future__ import annotations
import re
import random
from abc import ABC
import io
import typing
from typing import *


class B64:

    B64T_STRING: str = None
    B64T_ARRAY: typing.List[str] = None

    @staticmethod
    def initialize_fields() -> None:
        B64.B64T_STRING: str = (
            "./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
        )

        B64.B64T_ARRAY: typing.List[str] = list(B64T_STRING)

    @staticmethod
    def getRandomSalt(num: int, random: random.Random) -> str:
        saltString = ""
        for i in range(1, num + 1):
            saltString += B64T_STRING[random.randint(0, len(B64T_STRING) - 1)]
        return saltString

    @staticmethod
    def b64from24bit(b2: int, b1: int, b0: int, outLen: int, buffer: str) -> None:
        w = ((b2 << 16) & 0x00FFFFFF) | ((b1 << 8) & 0x00FFFF) | (b0 & 0xFF)
        n = outLen
        while n > 0:
            buffer += B64T_ARRAY[w & 0x3F]
            w >>= 6
            n -= 1


B64.initialize_fields()
